test/test_matmul_cores/test_trig.cc

```diff
@@ -0,0 +1,645 @@
+{
+    protected:
+    virtual void SetUp(){};
+
+    virtual void TearDown(){};
+
+    inline static std::vector<uint32_t> keys {42};
+
+    // Helper function for generating a row-permutation matrix
+    // in ColMajor format
+    //NOTE: Need to deallocate memory explicitly
+    template <typename T>
+    static T* generate_row_permutation_row_major(int size, const std::vector<int>& final_row_order) {
+        // Allocate memory for a size x size matrix and initialize with zeros
+        T* perm_matrix = new T[size * size]();
+
+        // Fill in the matrix with '1's at positions corresponding to the final row order
+        for (int i = 0; i < size; ++i) {
+            int final_row = final_row_order[i];
+            perm_matrix[i * size + final_row] = 1.0;  // Set the '1' in the permuted position (RowMajor)
+        }
+
+        return perm_matrix;
+    }
+
+    // Helper function for explicitly generating a Hadamard matrix
+    // (Note that ColMajor and RowMajor storage is identical for `H`)
+    template <typename T>
+    static std::vector<T> generate_hadamard(int64_t log_n) {
+        int64_t size = 1 << log_n;  // size = 2^n
+        std::vector<std::vector<T>> H(size, std::vector<T>(size, 1));  // Initialize H_1
+
+        // Sylvester's construction: recursively build the matrix
+        for (int n = 1; n <= log_n; ++n) {
+            T curr_size = 1 << n;  // Current size of the matrix is 2^n
+            for (int i = 0; i < curr_size / 2; ++i) {
+                for (int j = 0; j < curr_size / 2; ++j) {
+                    // Fill the bottom-left and bottom-right quadrants
+                    H[i + curr_size / 2][j] = H[i][j];       // Copy the top-left quadrant to bottom-left
+                    H[i][j + curr_size / 2] = H[i][j];       // Copy the top-left quadrant to top-right
+                    H[i + curr_size / 2][j + curr_size / 2] = -H[i][j]; // Fill bottom-right with negative values
+                }
+            }
+        }
+
+        // Flatten into a vector in ColMajor order
+        std::vector<T> H_flat(size * size);
+
+        for (int col = 0; col < size; ++col) {
+            for (int row = 0; row < size; ++row) {
+                H_flat[col * size + row] = H[row][col];
+            }
+        }
+
+        return H_flat;
+    }
+
+    template <typename T>
+    static std::vector<T> generate_random_vector(int size, T lower_bound, T upper_bound) {
+        // Create a random device and seed the random number generator
+        // std::random_device rd;
+        std::mt19937 gen(42);
+
+        // Define the distribution range for the random Ts
+        std::uniform_real_distribution<> dist(lower_bound, upper_bound);
+
+        // Create a vector of the specified size
+        std::vector<T> random_vector(size);
+
+        // Generate random doubles and fill the vector
+        for (int i = 0; i < size; ++i) {
+            random_vector[i] = dist(gen);
+        }
+
+        return random_vector;
```

Comment:
Don't use STL for random number generation in these contexts. Get random matrices from:

https://github.com/BallisticLA/RandBLAS/blob/main/test/test_matmul_cores/linop_common.hh